core/runtime.rs

```diff
@@ -1549,34 +1554,73 @@ impl JsRuntime {
   }
 
   fn drain_macrotasks(&mut self) -> Result<(), Error> {
-    let js_macrotask_cb_handle =
-      match &Self::state(self.v8_isolate()).borrow().js_macrotask_cb {
-        Some(handle) => handle.clone(),
-        None => return Ok(()),
-      };
+    let state = Self::state(self.v8_isolate());
+
+    if state.borrow().js_macrotask_cbs.is_empty() {
+      return Ok(());
+    }
 
+    let js_macrotask_cb_handles = state.borrow().js_macrotask_cbs.clone();
     let scope = &mut self.handle_scope();
-    let js_macrotask_cb = js_macrotask_cb_handle.open(scope);
 
-    // Repeatedly invoke macrotask callback until it returns true (done),
-    // such that ready microtasks would be automatically run before
-    // next macrotask is processed.
-    let tc_scope = &mut v8::TryCatch::new(scope);
-    let this = v8::undefined(tc_scope).into();
-    loop {
-      let is_done = js_macrotask_cb.call(tc_scope, this, &[]);
+    for js_macrotask_cb_handle in js_macrotask_cb_handles {
+      let js_macrotask_cb = js_macrotask_cb_handle.open(scope);
 
-      if let Some(exception) = tc_scope.exception() {
-        return exception_to_err_result(tc_scope, exception, false);
-      }
+      // Repeatedly invoke macrotask callback until it returns true (done),
+      // such that ready microtasks would be automatically run before
+      // next macrotask is processed.
+      let tc_scope = &mut v8::TryCatch::new(scope);
+      let this = v8::undefined(tc_scope).into();
+      loop {
+        let is_done = js_macrotask_cb.call(tc_scope, this, &[]);
+
+        if let Some(exception) = tc_scope.exception() {
+          return exception_to_err_result(tc_scope, exception, false);
+        }
+
+        if tc_scope.has_terminated() || tc_scope.is_execution_terminating() {
+          return Ok(());
+        }
 
-      if tc_scope.has_terminated() || tc_scope.is_execution_terminating() {
-        break;
+        let is_done = is_done.unwrap();
+        if is_done.is_true() {
+          break;
+        }
       }
+    }
+
+    Ok(())
+  }
 
-      let is_done = is_done.unwrap();
-      if is_done.is_true() {
-        break;
+  fn drain_nexttick(&mut self) -> Result<(), Error> {
+    let state = Self::state(self.v8_isolate());
+
+    if state.borrow().js_nexttick_cbs.is_empty() {
+      return Ok(());
+    }
+
+    if !state.borrow().has_tick_scheduled {
+      let scope = &mut self.handle_scope();
+      scope.perform_microtask_checkpoint();
+    }
+
+    // TODO(bartlomieju): Node also checks for absence of "rejection_to_warn"
+    if !state.borrow().has_tick_scheduled {
+      return Ok(());
+    }
+
+    let js_nexttick_cb_handles = state.borrow().js_nexttick_cbs.clone();
```

Comment:
Non-blocking observation: `state.borrow()` gets called a lot. It's more efficient to borrow once and drop when necessary.

It's possible it doesn't matter in practice because the compiler "sees through" and elides borrow() calls.

Reply:
Please see comment above about risk of running into `BorrowError`